<a href="https://colab.research.google.com/github/rsys-adsingh/Machine-Learning/blob/main/study/Chapter_4.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Topic 1: Types of Machine Learning

Machine Learning algorithms are classified based on how they learn from data.

### **The Hierarchy**
```text
Machine Learning
│
├── Supervised Learning (Learning with a Teacher)
├── Unsupervised Learning (Self-Discovery)
├── Semi-Supervised Learning (Partial Guidance)
├── Self-Supervised Learning (Self-Labeling)
└── Reinforcement Learning (Trial and Error)
```

> **Core Difference:** The biggest differentiator is **"Does the algorithm know the correct answer while learning?"**

## Topic 1.1: Supervised Learning

### 1. Definition
Supervised Learning is a type of Machine Learning where the model learns from **labeled data**.

**Simple Formula:**
$$\text{Input} + \text{Correct Answer} \rightarrow \text{Learn} \rightarrow \text{Predict Future Answers}$$

### 2. Intuition
Imagine a teacher teaching a student:
1. **Teacher gives:** A Question + The Correct Answer.
2. **Student learns:** The pattern between them.
3. **Test:** The student is given a new Question and must predict the Answer.

### 3. Mathematical View
Supervised learning tries to learn a mapping function:
$$y = f(X)$$

*   **$X$**: Features (Inputs like Age, Income, Square Feet).
*   **$y$**: Labels (Correct answers like Survived, Price, Spam).
*   **$f$**: The "Brain" (Model) that maps inputs to outputs.

### 4. Types of Supervised Learning
There are two main branches based on the nature of the output ($y$):

#### **A. Regression (Predicting a Number)**
Predicts a **continuous numerical value**.
*   **Example:** Predicting House Price based on Area.

#### **B. Classification (Predicting a Category)**
Predicts a **discrete label or class**.
*   **Example:** Predicting if an email is "Spam" or "Not Spam".

In [1]:
import pandas as pd
from sklearn.linear_model import LinearRegression, LogisticRegression
import numpy as np

# --- 1. REGRESSION EXAMPLE (Simple House Price) ---
# X = Area (sq ft), y = Price ($)
data_reg = {
    'Area_sqft': [1000, 1500, 2000, 2500],
    'Price_USD': [200000, 300000, 400000, 500000]
}
df_reg = pd.DataFrame(data_reg)

# Initialize and Train
reg_model = LinearRegression()
reg_model.fit(df_reg[['Area_sqft']], df_reg['Price_USD'])

# Predict for 1800 sqft
pred_price = reg_model.predict([[1800]])
print(f"Regression (Price for 1800 sqft): ${pred_price[0]:.2f}")

# --- 2. CLASSIFICATION EXAMPLE (Simple Spam Filter) ---
# X = Number of symbols (!!!), y = 1 (Spam) / 0 (Not Spam)
data_clf = {
    'Symbols': [[1], [10], [2], [15]],
    'Is_Spam': [0, 1, 0, 1]
}

# Initialize and Train
clf_model = LogisticRegression()
clf_model.fit(data_clf['Symbols'], data_clf['Is_Spam'])

# Predict for a message with 12 symbols
pred_spam = clf_model.predict([[12]])
print(f"Classification (12 symbols is Spam?): {'Yes' if pred_spam[0] == 1 else 'No'}")

Regression (Price for 1800 sqft): $360000.00
Classification (12 symbols is Spam?): Yes


/usr/local/lib/python3.12/dist-packages/sklearn/utils/validation.py:2739: UserWarning: X does not have valid feature names, but LinearRegression was fitted with feature names
  warnings.warn(


### 5. Summary Table

| Feature | Regression | Classification |
| :--- | :--- | :--- |
| **Output Type** | Continuous (Number) | Discrete (Category) |
| **Question** | "How much?" | "Which one?" |
| **Example** | House Price, Temperature | Spam/Not Spam, Male/Female |
| **Goal** | Find a line/curve | Find a boundary |

### 6. Pros & Cons

**Advantages:**
*   ✅ **High Accuracy:** You tell the model exactly what is right.
*   ✅ **Easy Evaluation:** You can check how close the prediction was to the real label.

**Limitations:**
*   ❌ **Expensive:** Labeling data requires human effort.
*   ❌ **Bias:** If the labels are wrong, the model will learn wrong patterns.

## Topic 1.2: Unsupervised Learning 🔍

### 1. Definition
Unsupervised Learning is a type of Machine Learning where the model learns from **unlabeled data**.

**Simple Formula:**
$$\text{Input Data} \rightarrow \text{No Correct Answer} \rightarrow \text{Find Hidden Patterns}$$

### 2. Intuition
Imagine entering a room full of people. Nobody tells you who belongs together. You naturally group them based on Age, Profession, or Language. ML does the same by finding **Similarity**.

### 3. Mathematical View
Unlike supervised learning, there is no target $y$. Instead of learning $y = f(X)$, it tries to discover the **Structure ($S$)** of the data:
$$S = \text{pattern}(X)$$
It often uses **Distance** (like Euclidean distance) to decide if points are similar:
$$d = \sqrt{(x_2 - x_1)^2 + (y_2 - y_1)^2}$$

### 4. Main Problems in Unsupervised Learning
```text
Unsupervised Learning
│
├── Clustering (Grouping similar items)
├── Dimensionality Reduction (Simplifying features)
└── Association Rule Mining (Finding common pairs)
```

### A. Clustering
**Definition:** Grouping similar observations together.
*   **Example:** Segmenting mall customers into "VIP" and "Budget" shoppers.
*   **Algorithm:** K-Means

In [2]:
import pandas as pd
from sklearn.cluster import KMeans

# --- CLUSTERING EXAMPLE (Mall Customers) ---
# Features: [Annual Income ($), Spending Score (1-100)]
data_mall = {
    'Income': [15, 16, 80, 85, 40],
    'Spending': [80, 85, 20, 15, 50]
}
df_mall = pd.DataFrame(data_mall)

# Initialize KMeans to find 3 groups
kmeans = KMeans(n_clusters=3, random_state=42)
kmeans.fit(df_mall)

# Show results
df_mall['Group'] = kmeans.labels_
display(df_mall)

,Income,Spending,Group
0,15,80,0
1,16,85,0
2,80,20,1
3,85,15,1
4,40,50,2


### B. Dimensionality Reduction
**Definition:** Reduce the number of features while preserving the core information.
*   **Example:** Reducing 10 features to 2 so we can plot them on a graph.
*   **Algorithm:** PCA (Principal Component Analysis)

In [3]:
from sklearn.decomposition import PCA
import numpy as np

# --- DIMENSIONALITY REDUCTION EXAMPLE ---
# Original Data: 3 features (e.g., Height, Weight, Age)
X_complex = np.array([[170, 70, 30], [160, 60, 25], [180, 80, 35]])

# Reduce from 3D to 2D
pca = PCA(n_components=2)
X_simple = pca.fit_transform(X_complex)

print("Original Shape:", X_complex.shape)
print("Reduced Shape:", X_simple.shape)
print("\nReduced Data (2 features):\n", X_simple)

Original Shape: (3, 3)
Reduced Shape: (3, 2)

Reduced Data (2 features):
 [[-0.00000000e+00  1.87244452e-15]
 [-1.50000000e+01  2.07883101e-31]
 [ 1.50000000e+01  2.07883101e-31]]


### 5. Pros & Cons

**Advantages:**
*   ✅ **No labels required:** Useful for raw data.
*   ✅ **Discovery:** Finds patterns humans might miss.

**Limitations:**
*   ❌ **Hard to Evaluate:** There is no 'correct' score.
*   ❌ **Unpredictable:** Clusters might not always make business sense.

## Topic 1.3: Semi-Supervised Learning

### 1. Definition
Semi-Supervised Learning uses a **small amount of labeled data** and a **large amount of unlabeled data**.

**Simply put:**
Few Labels + Lots of Data

### 2. Example: Medical Images
*   **Data:** 10,000 X-rays.
*   **Constraint:** Only 500 are labeled by doctors (expensive).
*   **Solution:** The algorithm uses the 500 labels to learn basics and the 9,500 unlabeled images to understand general anatomy patterns.

### 3. Applications
*   Medical Diagnosis
*   Face Recognition
*   Speech Recognition

### 4. Pros & Cons
*   ✅ **Pros:** Reduces labeling cost; better than using only a small labeled dataset.
*   ❌ **Cons:** Performance depends heavily on the quality of those few labeled samples.

## Topic 1.4: Self-Supervised Learning ⭐⭐⭐

### 1. Definition
Self-Supervised Learning creates its own labels from the data. The data itself generates the learning task.

### 2. Example: Sentence Completion
*   **Input:** "I love Machine Learning"
*   **Task:** Hide one word -> "I love ____ Learning"
*   **Label:** The model predicts "Machine".
*   **Note:** No human labeled this; the sentence provided the answer!

### 3. Applications
*   **LLMs:** ChatGPT, BERT, GPT
*   **Vision:** CLIP, DINO

### 4. Advantages
*   ✅ No manual labeling required.
*   ✅ Can learn from billions of examples (the whole internet).

## Topic 1.5: Reinforcement Learning ⭐⭐⭐⭐⭐

### 1. Definition
Reinforcement Learning (RL) is where an **Agent** learns by interacting with an **Environment** and receiving **Rewards** or penalties.

**Simply put:**
Trial → Feedback → Improve

### 2. Intuition
Teaching a dog a trick:
*   **Correct trick:** Reward (Treat)
*   **Wrong trick:** No reward
*   Eventually, the dog learns the behavior that gets the most treats.

### 3. Mathematical View
At every step, the Agent observes **State (s)**, takes an **Action (a)**, and receives a **Reward (r)**.
**Goal:** Maximize the Total Reward over time.

### 4. Applications
*   Self-driving Cars
*   Robotics
*   Game AI (AlphaGo)
*   ChatGPT RLHF (Reinforcement Learning from Human Feedback)

## Complete Comparison ⭐⭐⭐⭐⭐

| Learning Type | Labels Available? | Goal | Example |
| :--- | :--- | :--- | :--- |
| **Supervised** | ✅ Yes | Predict labels | Titanic (Survival) |
| **Unsupervised** | ❌ No | Find patterns | Customer Segments |
| **Semi-Supervised** | 🌓 Partial | Improve with few labels | Medical X-rays |
| **Self-Supervised** | 🤖 Self-generated | Learn representations | GPT, BERT |
| **Reinforcement** | 🏆 Rewards | Maximize reward | Robotics, Games |